### DISCUSSION TITLE
How to refine mesh when current solve failed ?

### DISCUSSION THREAD

**Discussion by stonebeta** [idaholab/moose#18748] [Q&A General] [answered]
I want to refine mesh, when current solve failed, not just to cut the time step.
Is that easy to implement ？Which files might need to be modified?

**Comment by GiudGiud** [comment, marked as answer]
I think it would not be hard to implement. You would run a step of adaptivity. We might need a new execution flag for adaptivity though.

I would look at Transient.C in framework/executioners
Around line 413 you ll see the logic when a time step failed. Then you could add a run of adaptivity there `_problem.adaptMesh();` I think

**Reply by stonebeta** [reply]
Thank you for your reply. 
Now I am trying to set the new execution flag according to the values of an aux variable ( or an indicator ). But I do not know how to get the pointer to my aux variables. Which function should I use ? I find some functions in `FEProblemBase.h`:
 ```c++
virtual const MooseVariableFieldBase &
  getVariable(THREAD_ID tid,
              const std::string & var_name,
              Moose::VarKindType expected_var_type = Moose::VarKindType::VAR_ANY,
              Moose::VarFieldType expected_var_field_type =
                  Moose::VarFieldType::VAR_FIELD_ANY) const override;

virtual MooseVariableFieldBase & getActualFieldVariable(THREAD_ID tid,
                                                  const std::string & var_name) override;

virtual MooseVariable & getStandardVariable(THREAD_ID tid, const std::string & var_name) override;
```
I do not know what the `THREAD_ID tid` means. If I should use these functions, what should I do to the `THREAD_ID tid` ?

**Reply by GiudGiud** [reply]
Since you are using the value, you should just use the Coupleable interface. You can find the code for that in interfaces/Coupleable.C and you can see many examples in kernels, like CoupledForce for example. 

tid can be set to _tid, which is the current thread id (=number)

I'm not sure about using aux variables over indicators for adaptivity. Can you give more details about this?

**Reply by stonebeta** [reply]
I want to add an additional adaptivity step, if the aux variable value of an element exceed the threshold and this element can still be refined (value > threshold && level < max_h_level). I have tried ValueThresholdMarker before. But the change of the mesh seems to be a step behind the change of the variable values.

**Reply by GiudGiud** [reply]
Ah it may be (it is) that auxckernels are not run when the solve failed. 
Can you base your mesh change on something else? Like a primary variable or a residual

**Reply by stonebeta** [reply]
Thank you for the reply.
I try to put this code into FixedPointSolve.C to check the variable value and the mesh level.
```c++
bool
FixedPointSolve::check_mesh_and_d()
{
  MooseVariable & _d_var = _problem.getStandardVariable(_tid,"d");
  MooseMesh & _mesh = _problem.mesh();
  for (dof_id_type i = 0; i < _mesh.nElem(); i++)
  {
    if (_mesh.elemPtr(i)->active()
      && (_mesh.elemPtr(i)->level() < _problem.adaptivity().getMaxHLevel())
      && (_d_var.getElementalValue(_mesh.elemPtr(i)) >= 1.0))
    {
      return false;
    }
  }
  return true;
}
```
It works when using only one processor. But it errors when using more processors.

```
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 106918 RUNNING AT node01
=   EXIT CODE: 139
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

The _tid here is only declared in the header file but not assigned value. I think this is the problem. Please help me, thanks.

**Reply by GiudGiud** [reply]
What does it return for _tid?

_tid is the thread id. Did you turn on threading or processes?

**Reply by stonebeta** [reply]
_tid is 0. I didn't do anything, just declared _tid.

**Reply by stonebeta** [reply]
`mpiexec -n 1 ./myapp-opt -i input.i` or `./myapp-opt -i input.i` is ok.
`mpiexec -n 4 ./myapp-opt -i input.i` would be wrong.

**Reply by GiudGiud** [reply]
ok tid seems fine. 
Let's get a backtrace on this error. Please compile in debug mode (`METHOD=dbg make`) then run in a debugger, one way is like this:
`mpirun -n 4 gdb --ex run --ex bt --args ./myapp-opt -i input.i`

**Reply by stonebeta** [reply]
Thread 1 "myapp-opt" received signal SIGSEGV, Segmentation fault.
0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#0  0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#1  0x00007ffff5372d03 in MooseVariableData<double>::getElementalValue(libMesh::Elem const*, Moose::SolutionState, unsigned int) const ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#2  0x00007ffff52f5a0c in FixedPointSolve::check_mesh_and_d() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#3  0x00007ffff5319d4e in FixedPointSolve::solve() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#4  0x00007ffff5f7687e in TimeStepper::step() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#5  0x00007ffff52f981e in Transient::takeStep(double) ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#6  0x00007ffff52f600e in Transient::execute() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#7  0x00007ffff60d420a in MooseApp::executeExecutioner() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#8  0x00007ffff60d8939 in MooseApp::run() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#9  0x0000555555557153 in main ()

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#0  0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#1  0x00007ffff5372d03 in MooseVariableData<double>::getElementalValue(libMesh::Elem const*, Moose::SolutionState, unsigned int) const ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#2  0x00007ffff52f5a0c in FixedPointSolve::check_mesh_and_d() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#3  0x00007ffff5319d4e in FixedPointSolve::solve() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#4  0x00007ffff5f7687e in TimeStepper::step() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#5  0x00007ffff52f981e in Transient::takeStep(double) ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#6  0x00007ffff52f600e in Transient::execute() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#7  0x00007ffff60d420a in MooseApp::executeExecutioner() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#8  0x00007ffff60d8939 in MooseApp::run() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#0  0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#0  0x00007ffff614d20a in libMesh::PetscVector<double>::operator()(unsigned long) const () from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#1  0x00007ffff5372d03 in MooseVariableData<double>::getElementalValue(libMesh::Elem const*, Moose::SolutionState, unsigned int) const ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#1  0x00007ffff5372d03 in MooseVariableData<double>::getElementalValue(libMesh::Elem const*, Moose::SolutionState, unsigned int) const ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#9  0x0000555555557153 in main ()
#2  0x00007ffff52f5a0c in FixedPointSolve::check_mesh_and_d() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#2  0x00007ffff52f5a0c in FixedPointSolve::check_mesh_and_d() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#3  0x00007ffff5319d4e in FixedPointSolve::solve() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#3  0x00007ffff5319d4e in FixedPointSolve::solve() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#4  0x00007ffff5f7687e in TimeStepper::step() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#4  0x00007ffff5f7687e in TimeStepper::step() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#5  0x00007ffff52f981e in Transient::takeStep(double) ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#5  0x00007ffff52f981e in Transient::takeStep(double) ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#6  0x00007ffff52f600e in Transient::execute() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#6  0x00007ffff52f600e in Transient::execute() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#7  0x00007ffff60d420a in MooseApp::executeExecutioner() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#7  0x00007ffff60d420a in MooseApp::executeExecutioner() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#8  0x00007ffff60d8939 in MooseApp::run() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#8  0x00007ffff60d8939 in MooseApp::run() ()
   from /home/redslime/projects/moose/framework/libmoose-opt.so.0
#9  0x0000555555557153 in main ()
#9  0x0000555555557153 in main ()

**Reply by GiudGiud** [reply]
Hello

It's better to generate back traces using a debug executable, that way you get the line number. 
It ll also print the arguments passed in every function, it's just a lot more helpful.

One part I'm not sure about is this _mesh.elemPtr(i)->level() < _problem.adaptivity().getMaxHLevel()
which elements are you trying to exclude here? Are you trying to limit the number of refinements?

Another is that you are looping over all elements (nElems) but you are domain decomposed. So this needs to be restricted to the elements on the domain. See these pages for the code documentation:
https://mooseframework.inl.gov/docs/doxygen/moose/classMooseMesh.html#ae5c470c83c346fbf9970755391387777
https://libmesh.github.io/doxygen/classlibMesh_1_1MeshBase.html#a190e556d3be87f5ccd8d4fc98d1779d3

So you should use this routine instead, getActiveLocalElementRange(), it's better practice too to loop directly on the elements. 

Guillaume

**Reply by stonebeta** [reply]
Thank you for your patient reply. I modified my code:
```c
bool
FixedPointSolve::check_mesh_and_d()
{
  MooseVariable & _d_var = _problem.getStandardVariable(_tid,"d");
  MooseMesh & _mesh = _problem.mesh();
  for (auto && element:*(_mesh.getActiveLocalElementRange()))
  {
    if ((element->level() < _problem.adaptivity().getMaxHLevel())
      && (_d_var.getElementalValue(element) >= _max_d_in_coarsened_mesh))
    {
      return false;
    }
  }
  return true;
}
```
It is ok when check_mesh_and_d() returns true.
But it will stuck when check_mesh_and_d() returns false, like this:
```
Starting Transfers on TIMESTEP_END From MultiApps
Beginning MultiAppCopyTransfer from_d
Finished MultiAppCopyTransfer from_d
Transfers on TIMESTEP_END Are Finished

Fixed point residual norm after TIMESTEP_END MultiApps: 8.321663e-11

 0 Picard |R| = 1.577468e-03
 1 Picard |R| = 5.581505e-06
 2 Picard |R| = 1.088676e-06
 3 Picard |R| = 2.252272e-07
 4 Picard |R| = 4.658259e-08
 5 Picard |R| = 9.598667e-09
 6 Picard |R| = 1.973526e-09
 7 Picard |R| = 4.053446e-10
 8 Picard |R| = 8.321663e-11
Fixed point convergence reason: CONVERGED_ABS
Executing..........
```
Is that because check_mesh_and_d() may return different values on the decomposed domain ?

**Reply by GiudGiud** [reply]
That's pretty likely yes!
One way to get around this is to reduce an integer across all ranks using _comm.sum(some_integer)
If the integer is above 0, then do the work. If all ranks are at 0, skip adaptivity

**Reply by GiudGiud** [reply]
_comm or _communicator are likely a class attribute. If not you can get it from the mesh btw _mesh->comm()
